// @appland/appmap/built/cmds/inventory/analyzers/collectFindings.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const readIndexFile_1 = __importDefault(require("../readIndexFile"));
function collectFindings(findings) {
    const modifiedDate = (finding) => finding.eventsModifiedDate || finding.scopeModifiedDate;
    return async (appmap) => {
        const scanResults = (await (0, readIndexFile_1.default)(appmap, 'appmap-findings.json'));
        if (!scanResults)
            return;
        for (const finding of scanResults.findings) {
            const mDate = modifiedDate(finding);
            if (mDate)
                findings.push({
                    appmap: appmap.slice(0, -'.appmap.json'.length),
                    ruleId: finding.ruleId,
                    impactDomain: finding.impactDomain,
                    modifiedDate: new Date(mDate),
                    hash_v2: finding.hash_v2,
                });
        }
    };
}
exports.default = collectFindings;
