// @appland/appmap/built/cmds/compare/ui.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.sanitizeRevision = exports.mutedStyle = exports.prominentStyle = exports.actionStyle = exports.confirm = exports.ask = void 0;
const chalk_1 = __importDefault(require("chalk"));
async function ask(rl, q) {
    return new Promise((resolve) => {
        rl.question(q, resolve);
    });
}
exports.ask = ask;
async function confirm(prompt, rl) {
    let response = '';
    while (!['y', 'n'].includes(response)) {
        response = await ask(rl, `${prompt} (y/n) `);
    }
    return response === 'y';
}
exports.confirm = confirm;
function actionStyle(message) {
    return chalk_1.default.bold(chalk_1.default.green(message));
}
exports.actionStyle = actionStyle;
function prominentStyle(message) {
    return chalk_1.default.bold(message);
}
exports.prominentStyle = prominentStyle;
function mutedStyle(message) {
    return chalk_1.default.dim(message);
}
exports.mutedStyle = mutedStyle;
function sanitizeRevision(revision) {
    return revision.replace(/[^a-zA-Z0-9_]/g, '_');
}
exports.sanitizeRevision = sanitizeRevision;
