// @appland/openapi/dist/appmap/propertiesParserV2.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const schemaInferrer_1 = require("../schemaInferrer");
const util_1 = require("../util");
const mergeReducer = (acc, obj) => (0, schemaInferrer_1.mergeType)(obj, acc) || acc || acc;
function parseProperty(paramProperty) {
    const type = (0, util_1.classNameToOpenAPIType)(paramProperty.class);
    if (type === undefined)
        return undefined;
    if (paramProperty.items) {
        return parseArrayProperties(paramProperty.items);
    }
    if (type === 'array') {
        // If we made it here, we're an array with no additional information.
        return { type: 'array', items: {} };
    }
    if (paramProperty.properties) {
        return parseObjectProperties(paramProperty.properties);
    }
    // `type` should be an expected value, but TypeScript doesn't know that. Maybe there's
    // a smarter cast we can do here, but it's beyond me at the moment.
    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-explicit-any
    return { type: type };
}
function parseArrayProperties(paramProperties) {
    const itemArray = paramProperties.map(parseProperty).filter(Boolean);
    if (itemArray.length === 0)
        return { type: 'array', items: {} };
    else if (itemArray.length === 1)
        return { type: 'array', items: itemArray[0] };
    const primitiveTypes = new Set(itemArray
        .filter((item) => item.type && item.type !== 'object' && item.type !== 'array')
        .map(({ type }) => type));
    const primitives = Array.from(primitiveTypes).map((type) => ({
        type,
    }));
    const mergedArrays = itemArray
        .filter((item) => item.type === 'array')
        .reduce(mergeReducer, { items: {} });
    const mergedObject = itemArray
        .filter((item) => item.type === 'object')
        .reduce(mergeReducer, {});
    const allMergedItems = [...primitives, mergedArrays, mergedObject].filter((item) => item && item.type);
    if (allMergedItems.length === 0) {
        return { type: 'array', items: {} };
    }
    if (allMergedItems.length === 1) {
        return { type: 'array', items: allMergedItems[0] };
    }
    return { type: 'array', items: { anyOf: allMergedItems } };
}
function parseObjectProperties(paramProperties) {
    const properties = paramProperties.filter(Boolean).reduce((memo, msgProperty) => {
        const property = parseProperty(msgProperty);
        if (property)
            memo[msgProperty.name] = property;
        return memo;
    }, {});
    return { type: 'object', properties };
}
class PropertiesParser {
    static canParse(example) {
        var _a;
        if (!example.properties)
            return true;
        const properties = [...example.properties, example];
        while (properties.length) {
            const property = properties.pop();
            if (!property)
                continue;
            const type = (0, util_1.classNameToOpenAPIType)(property.class);
            if (!type)
                continue;
            if (type === 'array' && property.items === undefined)
                return false;
            if (property.name === undefined)
                return true;
            if (property.class === undefined)
                return false;
            if (property.items !== undefined)
                return true;
            (_a = property.properties) === null || _a === void 0 ? void 0 : _a.forEach((p) => properties.push(p));
        }
        return false;
    }
    static parse(example) {
        return parseProperty(example);
    }
}
exports.default = PropertiesParser;
