// @appland/appmap/built/cmds/agentInstaller/NodeNoInstaller.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const node_path_1 = require("node:path");
const utils_1 = require("../../utils");
const agentInstaller_1 = __importDefault(require("./agentInstaller"));
class NodeNoInstaller extends agentInstaller_1.default {
    constructor(path) {
        super('Node.js', path);
        this.buildFile = 'package.json';
        this.documentation = 'https://appmap.io/docs/reference/appmap-node';
        this.language = 'javascript';
        this.appmap_dir = 'tmp/appmap';
        this.isNoop = true;
    }
    installAgent(ui) {
        ui.message('No AppMap installation needed for Node.js projects.');
        ui.message('You can use it directly by prepending `npx appmap-node` to your commands.');
        ui.message(`For more information, see ${this.documentation}`);
    }
    available() {
        return (0, utils_1.exists)((0, node_path_1.join)(this.path, this.buildFile));
    }
    environment() {
        return { 'Node version': process.version };
    }
    checkConfigCommand(ui) { }
    validateAgentCommand() { }
    initCommand() { }
    verifyCommand() { }
}
exports.default = NodeNoInstaller;
