// @appland/client/dist/src/usage.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const configuration_1 = require("./configuration");
const makeRequest_1 = __importDefault(require("./makeRequest"));
class Usage {
    static async get() {
        const response = await (0, makeRequest_1.default)({
            service: configuration_1.ServiceEndpoint.ServiceApi,
            path: this.usagePath,
        });
        const responseJson = JSON.parse(response.body.toString('utf-8'));
        if (!response.ok) {
            if ('message' in responseJson)
                throw new Error(responseJson.message);
            throw new Error(`Got unexpected status code ${response.statusCode}`);
        }
        return responseJson;
    }
    static async update(dto) {
        // Consider this to be a no-op if there's nothing to report.
        if (!dto.events && !dto.appmaps)
            return;
        await (0, makeRequest_1.default)({
            service: configuration_1.ServiceEndpoint.ServiceApi,
            path: this.usagePath,
            method: 'POST',
            body: JSON.stringify(dto),
            retry: false,
        });
    }
}
exports.default = Usage;
Usage.usagePath = 'v1/usage';
