// @appland/appmap/built/fingerprint/canonicalize/exceptions.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const unique_1 = __importDefault(require("./unique"));
function formatException(event, exception) {
    const result = {
        class: exception.class,
        message: exception.message,
        stack: [],
    };
    let { path, lineno } = exception;
    if (path) {
        result.location = [path, lineno].filter(Boolean).join(':');
        result.stack.push(result.location);
    }
    let ancestor = event;
    while (ancestor) {
        if (ancestor.codeObject.location) {
            result.stack.push(ancestor.codeObject.location);
        }
        ancestor = ancestor.parent;
    }
    return result;
}
class Canonicalize extends unique_1.default {
    functionCall(event) {
        const { exceptions } = event;
        if (!exceptions || exceptions.length === 0)
            return;
        return exceptions.map((exception) => formatException(event, exception)).filter(Boolean);
    }
}
module.exports = (appmap) => new Canonicalize(appmap).execute();
