// @appland/appmap/built/cmds/record/state/initial.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.createState = void 0;
const userInteraction_1 = __importDefault(require("../../userInteraction"));
async function createState(method) {
    return (await (_a = `./record_${method}`, Promise.resolve().then(() => __importStar(require(_a))))).default;
}
exports.createState = createState;
async function initial(recordContext) {
    const choices = {
        'test cases': 'test',
        'remote recording': 'remote',
    };
    const { method: methodName } = await userInteraction_1.default.prompt({
        name: 'method',
        type: 'list',
        message: 'Choose recording method:',
        choices: Object.keys(choices),
    });
    const method = choices[methodName];
    recordContext.recordMethod = method;
    return createState.bind(null, method);
}
exports.default = initial;
