// @appland/appmap/built/lib/eventAggregator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MaxMSBetween = void 0;
exports.MaxMSBetween = 10 * 1000;
class EventAggregator {
    constructor(callback, maxMsBetween = exports.MaxMSBetween) {
        this.callback = callback;
        this.maxMsBetween = maxMsBetween;
        this.pending = [];
        process.on('beforeExit', () => this.dispose());
    }
    push(emitter, event, args) {
        this.pending.push({ emitter, event, args });
        this.refresh();
    }
    refresh() {
        if (this.timeout)
            clearTimeout(this.timeout);
        this.timeout = setTimeout(this.emitPending.bind(this), this.maxMsBetween);
    }
    async emitPending() {
        this.timeout = undefined;
        if (this.pending.length) {
            const pendingEvents = this.pending.splice(0, this.pending.length);
            await this.callback(pendingEvents);
        }
    }
    attach(emitter, event) {
        emitter.on(event, (...args) => this.push(emitter, event, args));
    }
    async dispose() {
        if (this.timeout) {
            clearTimeout(this.timeout);
            await this.emitPending();
        }
    }
}
exports.default = EventAggregator;
