// @appland/appmap/built/cmds/archive/generateSequenceDiagrams.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateSequenceDiagrams = void 0;
const processAppMapDir_1 = __importDefault(require("../../lib/processAppMapDir"));
const utils_1 = require("../../utils");
async function generateSequenceDiagrams(workerPool, maxAppMapSizeInBytes, compareFilter, oversizedAppMaps, appMapDir) {
    const specOptions = {
        loops: true,
    };
    const task = (file) => ({
        name: 'sequence-diagram',
        verbose: (0, utils_1.verbose)(),
        appmapFile: file,
        maxSize: maxAppMapSizeInBytes,
        compareFilter,
        specOptions,
    });
    const startTime = new Date().getTime();
    const result = await (0, processAppMapDir_1.default)('Generating sequence diagrams', workerPool, task, appMapDir);
    const elapsed = new Date().getTime() - startTime;
    console.log(`Generated ${result.numProcessed} sequence diagrams in ${elapsed}ms`);
    for (const file of result.oversized)
        oversizedAppMaps.add(file);
}
exports.generateSequenceDiagrams = generateSequenceDiagrams;
