// @appland/openapi/dist/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SecuritySchemes = exports.rpcRequestForEvent = exports.forClientRequest = exports.Model = exports.verbose = exports.classNameToOpenAPIType = exports.parseHTTPServerRequests = void 0;
const model_1 = __importDefault(require("./model"));
exports.Model = model_1.default;
const securitySchemes_1 = __importDefault(require("./securitySchemes"));
exports.SecuritySchemes = securitySchemes_1.default;
const rpcRequest_1 = require("./rpcRequest");
Object.defineProperty(exports, "rpcRequestForEvent", { enumerable: true, get: function () { return rpcRequest_1.rpcRequestForEvent; } });
var parseHTTPServerRequests_1 = require("./parseHTTPServerRequests");
Object.defineProperty(exports, "parseHTTPServerRequests", { enumerable: true, get: function () { return __importDefault(parseHTTPServerRequests_1).default; } });
var util_1 = require("./util");
Object.defineProperty(exports, "classNameToOpenAPIType", { enumerable: true, get: function () { return util_1.classNameToOpenAPIType; } });
Object.defineProperty(exports, "verbose", { enumerable: true, get: function () { return util_1.verbose; } });
const forClientRequest = (event) => {
    const rpcRequest = (0, rpcRequest_1.rpcRequestForEvent)(event);
    if (!rpcRequest) {
        return;
    }
    const securitySchemes = new securitySchemes_1.default();
    const model = new model_1.default();
    if (rpcRequest.requestHeaders) {
        const authorizationHeader = (0, rpcRequest_1.headerValue)(rpcRequest.requestHeaders, 'authorization');
        if (authorizationHeader)
            securitySchemes.addAuthorizationHeader(authorizationHeader);
    }
    model.addRpcRequest(rpcRequest);
    return {
        warnings: model.collectWarnings(),
        paths: model.openapi(),
        securitySchemes: securitySchemes.openapi(),
    };
};
exports.forClientRequest = forClientRequest;
