// @appland/openapi/dist/method.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const response_1 = __importDefault(require("./response"));
const rpcRequest_1 = require("./rpcRequest");
const objectSchema_1 = __importDefault(require("./objectSchema"));
const util_1 = require("./util");
const Warnings_1 = __importDefault(require("./Warnings"));
const bodyParamMethods = new Set(['options', 'put', 'post', 'patch']);
function openapiIn(rpcRequest, name) {
    const pathInfo = rpcRequest.requestPath;
    // Trim format info from Rails paths, e.g. /foo/bar(.:format)
    const tokens = pathInfo.split('/').map((token) => token.split('(')[0]);
    // Recognize Rails-style normalized paths /org/:org_id and OpenAPI-style paths /org/{org_id}
    if (tokens.includes(`:${name}`) || tokens.includes(`{${name}}`)) {
        return 'path';
    }
    return 'query';
}
class Method {
    constructor() {
        this.rpcRequests = [];
        this.responses = {};
        this.warnings = new Warnings_1.default();
    }
    get containedWarningProviders() {
        return Object.values(this.responses);
    }
    get summaries() {
        return this.rpcRequests
            .map((request) => (0, rpcRequest_1.headerValue)(request.responseHeaders, 'x-openapi-summary'))
            .filter(Boolean);
    }
    openapi() {
        const responseByStatus = Object.keys(this.responses)
            .sort()
            .reduce((memo, status) => {
            // eslint-disable-next-line no-param-reassign
            memo[status] = this.responses[status].openapi();
            return memo;
        }, {});
        const schemata = {};
        const parameters = [];
        let securitySchemeId;
        const uniqueNames = new Set(['controller', 'action', '_method']);
        this.rpcRequests.forEach((rpcRequest) => {
            const messages = rpcRequest.parameters;
            if (!securitySchemeId) {
                const authorization = rpcRequest.requestHeaders['Authorization'];
                if (authorization) {
                    const scheme = (0, util_1.parseScheme)(authorization);
                    securitySchemeId = scheme.schemeId;
                }
            }
            let schema;
            if (rpcRequest.requestContentType) {
                const mimeType = rpcRequest.requestContentType.split(';')[0];
                if (!schemata[mimeType]) {
                    schemata[mimeType] = new objectSchema_1.default();
                }
                schema = schemata[mimeType];
            }
            messages.forEach((message) => {
                if (!message.name) {
                    return;
                }
                if (uniqueNames.has(message.name)) {
                    return;
                }
                uniqueNames.add(message.name);
                const inLocation = openapiIn(rpcRequest, message.name);
                if (inLocation !== 'path' &&
                    rpcRequest.requestMethod &&
                    bodyParamMethods.has(rpcRequest.requestMethod)) {
                    if (schema) {
                        schema.addExample(message);
                    }
                }
                else {
                    const schema = (0, util_1.messageToOpenAPISchema)(message);
                    if (!schema)
                        return;
                    const parameter = {
                        name: message.name,
                        in: inLocation,
                        schema,
                    };
                    if (parameter.in === 'path') {
                        parameter.required = true;
                    }
                    parameters.push(parameter);
                }
            });
        });
        const response = {
            responses: responseByStatus,
        };
        if (securitySchemeId) {
            const securityObj = {};
            securityObj[securitySchemeId] = [];
            response.security = [securityObj];
        }
        const mediaTypes = Object.keys(schemata)
            .sort()
            .map((contentType) => ({
            contentType,
            schema: schemata[contentType].schema(),
        }))
            .filter((entry) => entry.schema);
        if (mediaTypes.length > 0) {
            const content = mediaTypes.reduce((memo, entry) => {
                memo[entry.contentType] = { schema: entry.schema };
                return memo;
            }, {});
            response.requestBody = { content };
        }
        if (parameters.length > 0) {
            response.parameters = parameters.sort((a, b) => a.name.localeCompare(b.name));
        }
        const summaryFrequency = new Map();
        this.summaries.forEach((summary) => summaryFrequency.set(summary, (summaryFrequency.get(summary) || 0) + 1));
        if (summaryFrequency.size > 0) {
            response.summary = [...summaryFrequency.keys()].sort(
            // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
            (a, b) => summaryFrequency.get(b) - summaryFrequency.get(a))[0];
        }
        return response;
    }
    addRpcRequest(request) {
        const { status } = request;
        const statusError = () => {
            if (!status)
                return `Missing or undefined HTTP status code`;
            if (status < 100 || status >= 600)
                return `Invalid HTTP status code: ${status}`;
        };
        const error = statusError();
        if (error) {
            this.warnings.add((0, rpcRequest_1.routeId)(request), error);
            return;
        }
        if (!this.responses[status]) {
            this.responses[status] = new response_1.default(status);
        }
        this.rpcRequests.push(request);
        const responseObj = this.responses[status];
        responseObj.addRpcRequest(request);
    }
}
exports.default = Method;
