// @appland/openapi/dist/model.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = __importDefault(require("./path"));
const rpcRequest_1 = require("./rpcRequest");
const Warnings_1 = __importDefault(require("./Warnings"));
class Model {
    constructor() {
        this.paths = {};
        this.warnings = new Warnings_1.default();
    }
    collectWarnings() {
        const warnings = this.warnings.dup();
        warnings.merge(Object.values(this.paths));
        return warnings.warnings;
    }
    openapi() {
        const paths = Object.keys(this.paths)
            .sort()
            .reduce((memo, path) => {
            // eslint-disable-next-line no-param-reassign
            const pathItem = this.paths[path].openapi();
            if (Object.keys(pathItem).length) {
                memo[path] = this.paths[path].openapi();
            }
            return memo;
        }, {});
        return paths;
    }
    addRpcRequest(rpcRequest) {
        const path = Model.basePath(rpcRequest.requestPath);
        if (!path.startsWith('/')) {
            this.warnings.add((0, rpcRequest_1.routeId)(rpcRequest), `Path must start with '/'`);
            return;
        }
        if (!this.paths[path]) {
            this.paths[path] = new path_1.default();
        }
        const pathObj = this.paths[path];
        pathObj.addRpcRequest(rpcRequest);
    }
    static basePath(path) {
        const pathTokens = path.split('/').map((entry) => {
            if (entry.match(/^:(.*)/)) {
                // eslint-disable-next-line no-param-reassign
                entry = `{${entry.substring(1)}}`;
            }
            return entry;
        });
        if (pathTokens.length === 1 && path[0] === '') {
            pathTokens.push('');
        }
        return pathTokens.join('/');
    }
}
exports.default = Model;
