// @appland/openapi/dist/objectSchema.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const schemaInferrer_1 = __importDefault(require("./schemaInferrer"));
class ObjectSchema {
    constructor() {
        this.examples = [];
    }
    addExample(message) {
        const hasName = (p) => !!p.name;
        if (hasName(message)) {
            this.examples.push(message);
        }
    }
    get empty() {
        return this.examples.length > 0;
    }
    schema() {
        const schemata = {};
        this.examples
            .filter((message) => message.name)
            .sort((a, b) => a.name.localeCompare(b.name))
            .forEach((message) => {
            if (!schemata[message.name]) {
                schemata[message.name] = new schemaInferrer_1.default();
            }
            schemata[message.name].addExample(message);
        });
        if (Object.keys(schemata).length === 0) {
            return;
        }
        const properties = Object.keys(schemata).reduce((memo, name) => {
            const schema = schemata[name].openapi();
            if (schema)
                memo[name] = schema;
            return memo;
        }, {});
        return {
            type: 'object',
            properties,
        };
    }
}
exports.default = ObjectSchema;
