// @appland/openapi/dist/path.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const method_1 = __importDefault(require("./method"));
const Warnings_1 = __importDefault(require("./Warnings"));
class Path {
    constructor() {
        this.methods = {};
        this.warnings = new Warnings_1.default();
    }
    get containedWarningProviders() {
        return Object.values(this.methods);
    }
    openapi() {
        return Object.keys(this.methods)
            .sort()
            .reduce((memo, methodName) => {
            const method = this.methods[methodName];
            const methodOpenAPI = method.openapi();
            // Don't emit empty methods.
            if (Object.keys(methodOpenAPI.responses).length) {
                // eslint-disable-next-line no-param-reassign
                memo[methodName] = methodOpenAPI;
            }
            return memo;
        }, {});
    }
    addRpcRequest(rpcRequest) {
        const method = rpcRequest.requestMethod;
        if (!method) {
            return;
        }
        if (!this.methods[method]) {
            this.methods[method] = new method_1.default();
        }
        const methodObj = this.methods[method];
        methodObj.addRpcRequest(rpcRequest);
    }
}
exports.default = Path;
