// @appland/openapi/dist/response.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const statusCodes_1 = __importDefault(require("./statusCodes"));
const schemaInferrer_1 = __importDefault(require("./schemaInferrer"));
const Warnings_1 = __importDefault(require("./Warnings"));
class Response {
    constructor(statusCode) {
        this.statusCode = statusCode;
        this.rpcRequests = [];
        this.warnings = new Warnings_1.default();
    }
    get containedWarningProviders() {
        return [];
    }
    openapi() {
        // eslint-disable-next-line arrow-body-style
        const contentTypes = () => {
            return this.rpcRequests
                .map((rpcRequest) => rpcRequest.responseContentType)
                .filter((ct) => !!ct)
                .map((ct) => ct.split(';')[0]);
        };
        const schemata = {};
        this.rpcRequests.forEach((rpcRequest) => {
            const returnValue = rpcRequest.returnValue;
            if (!returnValue)
                return;
            if (rpcRequest.responseContentType) {
                const mimeType = rpcRequest.responseContentType.split(';')[0];
                if (!schemata[mimeType]) {
                    schemata[mimeType] = new schemaInferrer_1.default();
                }
                schemata[mimeType].addExample(returnValue);
            }
        });
        const content = [...new Set(contentTypes())].sort().reduce((memo, mimeType) => {
            memo[mimeType] = {};
            if (!schemata[mimeType])
                return memo;
            // eslint-disable-next-line no-param-reassign
            const schema = schemata[mimeType].openapi();
            if (schema)
                memo[mimeType].schema = schema;
            return memo;
        }, {});
        return { content, description: statusCodes_1.default[this.statusCode] };
    }
    addRpcRequest(rpcRequest) {
        this.rpcRequests.push(rpcRequest);
    }
}
exports.default = Response;
