// @appland/openapi/dist/rpcRequest.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.rpcRequestForEvent = exports.headerValue = exports.routeId = void 0;
const url_1 = require("url");
const util_1 = require("./util");
class ServerRPCRequest {
    constructor(event) {
        this.event = event;
    }
    get status() {
        return this.event.httpServerResponse.status;
    }
    get parameters() {
        return this.event.message || [];
    }
    get returnValue() {
        return this.event.returnValue;
    }
    get requestHeaders() {
        var _a;
        return ((_a = this.event.httpServerRequest) === null || _a === void 0 ? void 0 : _a.headers) || {};
    }
    get responseHeaders() {
        var _a;
        return ((_a = this.event.httpServerResponse) === null || _a === void 0 ? void 0 : _a.headers) || {};
    }
    get requestContentType() {
        return this.event.requestContentType;
    }
    get responseContentType() {
        return this.event.responseContentType;
    }
    get requestMethod() {
        var _a;
        return (_a = this.event.requestMethod) === null || _a === void 0 ? void 0 : _a.toLowerCase();
    }
    get requestPath() {
        return (0, util_1.ensureString)(this.event.httpServerRequest.normalized_path_info || this.event.httpServerRequest.path_info);
    }
}
class ClientRPCRequest {
    constructor(event) {
        this.event = event;
    }
    get status() {
        return this.event.httpClientResponse.status;
    }
    get parameters() {
        return this.event.parameters || [];
    }
    get returnValue() {
        return this.event.returnValue;
    }
    get requestHeaders() {
        var _a;
        return ((_a = this.event.httpClientRequest) === null || _a === void 0 ? void 0 : _a.headers) || {};
    }
    get responseHeaders() {
        var _a;
        return ((_a = this.event.httpClientResponse) === null || _a === void 0 ? void 0 : _a.headers) || {};
    }
    get requestContentType() {
        return this.event.requestContentType;
    }
    get responseContentType() {
        return this.event.responseContentType;
    }
    get requestMethod() {
        var _a;
        return (_a = this.event.requestMethod) === null || _a === void 0 ? void 0 : _a.toLowerCase();
    }
    get requestPath() {
        // TODO: Back-substitute query parameters into the URL.
        return new url_1.URL(this.event.httpClientRequest.url).pathname;
    }
}
function isServerEvent(event) {
    var _a;
    return (!!event.httpServerRequest &&
        !!event.httpServerResponse &&
        !!(event.httpServerRequest.normalized_path_info || ((_a = event.httpServerRequest) === null || _a === void 0 ? void 0 : _a.path_info)));
}
function isClientEvent(event) {
    return !!event.httpClientRequest && !!event.httpClientResponse && !!event.httpClientRequest.url;
}
function routeId(request) {
    return [request.requestMethod.toUpperCase(), request.requestPath].join(' ');
}
exports.routeId = routeId;
function headerValue(headers, name) {
    const nameLower = name.toLowerCase();
    const key = Object.keys(headers).find((header) => header.toLowerCase() === nameLower);
    if (key)
        return headers[key];
}
exports.headerValue = headerValue;
function rpcRequestForEvent(event) {
    if (isServerEvent(event))
        return new ServerRPCRequest(event);
    if (isClientEvent(event))
        return new ClientRPCRequest(event);
}
exports.rpcRequestForEvent = rpcRequestForEvent;
