// @appland/openapi/dist/schemaInferrer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.mergeType = exports.isAmbiguousArray = exports.mergeItems = void 0;
const util_1 = require("./util");
function mergeProperties(props1, props2) {
    const a = props1 || {};
    const b = props2 || {};
    const result = {};
    const props = new Set([...Object.keys(a), ...Object.keys(b)]);
    for (const key of props) {
        const merged = mergeType(a[key], b[key]);
        if (merged)
            result[key] = merged;
    }
    return result;
}
const uniqueObjects = (array) => {
    return [
        ...new Map(array.map((item) => [
            item.type || item.$ref,
            item,
        ])).values(),
    ];
};
function mergeItems(array1, array2) {
    const a = array1 || { type: 'array', items: {} };
    const b = array2 || { type: 'array', items: {} };
    if (a === b)
        return a;
    if (a.items.$ref || b.items.$ref) {
        console.warn('Cannot merge array items with references');
        return a;
    }
    const isAmbiguousArrayA = isAmbiguousArray(a);
    const isAmbiguousArrayB = isAmbiguousArray(b);
    if (isAmbiguousArrayA && isAmbiguousArrayB)
        return a;
    if (isAmbiguousArrayB && !isAmbiguousArrayA)
        return a;
    if (isAmbiguousArrayA && !isAmbiguousArrayB)
        return b;
    const { items: itemsA } = a;
    const { items: itemsB } = b;
    if (itemsA.type === 'object' && itemsB.type === 'object') {
        // TODO
        // Note that this means we do not support mixed object types in arrays.
        // They'll all be merged into a single type.
        return Object.assign(Object.assign({}, a), { items: Object.assign({}, mergeType(itemsA, itemsB)) });
    }
    if (itemsA.anyOf && itemsB.anyOf) {
        return Object.assign(Object.assign({}, a), { items: {
                anyOf: uniqueObjects(itemsA.anyOf.concat(itemsB.anyOf)),
            } });
    }
    if (itemsA.anyOf) {
        return Object.assign(Object.assign({}, a), { items: {
                anyOf: uniqueObjects(itemsA.anyOf.concat(itemsB)),
            } });
    }
    if (itemsB.anyOf) {
        return Object.assign(Object.assign({}, a), { items: {
                anyOf: uniqueObjects(itemsB.anyOf.concat(itemsA)),
            } });
    }
    if (itemsA.type === itemsB.type) {
        return a;
    }
    return Object.assign(Object.assign({}, a), { items: {
            anyOf: uniqueObjects([itemsA, itemsB]),
        } });
}
exports.mergeItems = mergeItems;
function isAmbiguousArray(schema) {
    if (schema.type === 'array') {
        return !schema.items || Object.keys(schema.items).length === 0;
    }
    return false;
}
exports.isAmbiguousArray = isAmbiguousArray;
// Merge type, items and properties of schema objects.
function mergeType(a, b) {
    if (!b)
        return a;
    if (!a)
        return b;
    if (b.type !== a.type)
        return a;
    if (a.type === 'array' && b.type === 'array') {
        const merged = mergeItems(a, b);
        if (merged)
            return merged;
    }
    if ('properties' in a && 'properties' in b) {
        return Object.assign(Object.assign({}, a), { properties: mergeProperties(a.properties, b.properties) });
    }
    return a;
}
exports.mergeType = mergeType;
class SchemaInferrer {
    constructor() {
        this.examples = [];
    }
    openapi() {
        const schemas = this.examples.map(util_1.messageToOpenAPISchema).filter(Boolean);
        if (schemas.length === 0)
            return;
        return schemas.reduce(mergeType);
    }
    addExample(value) {
        this.examples.push(value);
    }
}
exports.default = SchemaInferrer;
