// @appland/openapi/dist/util.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseScheme = exports.messageToOpenAPISchema = exports.ensureString = exports.classNameToOpenAPIType = exports.verbose = void 0;
const appmap_1 = require("./appmap");
const unrecognizedTypes = new Set();
function parseScheme(authorization) {
    const tokens = authorization.split(/\s/);
    if (tokens.length === 1) {
        return {
            schemeId: 'api_key',
            scheme: {
                type: 'apiKey',
                name: 'authorization',
                in: 'header',
            },
        };
    }
    const schemeId = tokens[0].toLowerCase();
    return {
        schemeId,
        scheme: {
            type: 'http',
            scheme: schemeId,
        },
    };
}
exports.parseScheme = parseScheme;
function classNameToOpenAPIType(className, options = {}) {
    if (!className || className === '')
        return;
    if (unrecognizedTypes.has(className))
        return 'object';
    const mapRubyType = (t) => {
        switch (t) {
            case 'array':
            case 'sequel::postgres::pgarray':
                return 'array';
            case 'hash':
            case 'sequel::postgres::jsonbhash':
            case 'activesupport::hashwithindifferentaccess':
                return 'object';
            case 'integer':
                return 'integer';
            case 'float':
            case 'numeric':
                return 'number';
            case 'trueclass':
            case 'falseclass':
                return 'boolean';
            case 'nilclass':
                return 'unknown';
            case 'string':
                return 'string';
        }
    };
    const mapPythonType = (t) => {
        if (!t.startsWith('builtins.')) {
            return;
        }
        switch (t.substring(9)) {
            case 'bool':
                return 'boolean';
            case 'dict':
                return 'object';
            case 'int':
                return 'integer';
            case 'list':
                return 'array';
            case 'str':
                return 'string';
            case 'nonetype':
                return 'unknown';
        }
    };
    const mapJavaType = (t) => {
        switch (t) {
            case 'java.lang.string':
                return 'string';
        }
    };
    const mapper = (t) => mapRubyType(t) || mapPythonType(t) || mapJavaType(t);
    const mapped = mapper(className.toLowerCase());
    if (!mapped) {
        if (options.strict) {
            return;
        }
        if (!unrecognizedTypes.has(className)) {
            if (verbose()) {
                console.warn(`Warning: Don't know how to map "${className}" to an OpenAPI type. You'll need to update the generated file.`);
            }
            unrecognizedTypes.add(className);
            return 'object';
        }
    }
    return mapped === 'unknown' ? undefined : mapped;
}
exports.classNameToOpenAPIType = classNameToOpenAPIType;
function messageToOpenAPISchema(example) {
    return (0, appmap_1.parse)(example);
}
exports.messageToOpenAPISchema = messageToOpenAPISchema;
function ensureString(value) {
    if (Array.isArray(value)) {
        return value.join('');
    }
    return value.toString();
}
exports.ensureString = ensureString;
let isVerbose = false;
function verbose(v) {
    if (v !== undefined) {
        isVerbose = v;
    }
    return isVerbose;
}
exports.verbose = verbose;
