// @appland/openapi/dist/Warnings.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class Warnings {
    constructor() {
        this.warnings = {};
    }
    dup() {
        const result = new Warnings();
        result.warnings = Object.assign({}, this.warnings);
        return result;
    }
    add(key, message) {
        if (!this.warnings[key])
            this.warnings[key] = [];
        this.warnings[key].push(message);
    }
    merge(warningProviders) {
        for (const warningProvider of warningProviders) {
            Object.entries(warningProvider.warnings.warnings).forEach(([key, messages]) => {
                messages.forEach((message) => {
                    this.add(key, message);
                });
            });
            this.merge(warningProvider.containedWarningProviders);
        }
    }
    static messages(warnings) {
        return Object.entries(warnings).reduce((memo, [key, messages]) => {
            memo.push(...messages.map((message) => `${key}: ${message}`));
            return memo;
        }, []);
    }
}
exports.default = Warnings;
