// @appland/navie/dist/context.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ContextV2 = void 0;
var ContextV2;
(function (ContextV2) {
    // Describes the different types of context data that is available.
    let ContextItemType;
    (function (ContextItemType) {
        // A summarized sequence diagram that describes the overall flow of the application.
        // The diagram is synthesized from a trace of the application's execution.
        ContextItemType["SequenceDiagram"] = "sequence-diagram";
        // A code snippet that is relevant to the context. The snippet is a few lines
        // from a source file that is related to the context.
        ContextItemType["CodeSnippet"] = "code-snippet";
        // A data request that was made by the application. This could be a database query,
        // an HTTP client request, or any other type of data request.
        ContextItemType["DataRequest"] = "data-request";
        // A help document that is relevant to the user's situation.
        ContextItemType["HelpDoc"] = "help-doc";
    })(ContextItemType = ContextV2.ContextItemType || (ContextV2.ContextItemType = {}));
    function isFileContextItem(contextItem) {
        return (contextItem.type === ContextItemType.CodeSnippet ||
            contextItem.type === ContextItemType.SequenceDiagram ||
            contextItem.type === ContextItemType.DataRequest);
    }
    ContextV2.isFileContextItem = isFileContextItem;
    let ContextLabelName;
    (function (ContextLabelName) {
        ContextLabelName["HelpWithAppMap"] = "help-with-appmap";
        ContextLabelName["Architecture"] = "architecture";
        ContextLabelName["Feature"] = "feature";
        ContextLabelName["Overview"] = "overview";
        ContextLabelName["Troubleshoot"] = "troubleshoot";
        ContextLabelName["Explain"] = "explain";
        ContextLabelName["Generate"] = "generate";
    })(ContextLabelName = ContextV2.ContextLabelName || (ContextV2.ContextLabelName = {}));
    let ContextLabelWeight;
    (function (ContextLabelWeight) {
        // The label is very relevant to the request.
        ContextLabelWeight["High"] = "high";
        // The label is somewhat relevant to the request.
        ContextLabelWeight["Medium"] = "medium";
    })(ContextLabelWeight = ContextV2.ContextLabelWeight || (ContextV2.ContextLabelWeight = {}));
})(ContextV2 = exports.ContextV2 || (exports.ContextV2 = {}));
