// @appland/navie/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.InteractionHistory = exports.Navie = exports.Help = exports.ProjectInfo = exports.ContextV2 = exports.Agents = exports.navie = exports.InteractionState = exports.applyContext = void 0;
/* eslint-disable import/prefer-default-export */
var apply_context_1 = require("./lib/apply-context");
Object.defineProperty(exports, "applyContext", { enumerable: true, get: function () { return __importDefault(apply_context_1).default; } });
var interaction_state_1 = require("./interaction-state");
Object.defineProperty(exports, "InteractionState", { enumerable: true, get: function () { return __importDefault(interaction_state_1).default; } });
var navie_1 = require("./navie");
Object.defineProperty(exports, "navie", { enumerable: true, get: function () { return __importDefault(navie_1).default; } });
var agent_1 = require("./agent");
Object.defineProperty(exports, "Agents", { enumerable: true, get: function () { return agent_1.AgentMode; } });
var context_1 = require("./context");
Object.defineProperty(exports, "ContextV2", { enumerable: true, get: function () { return context_1.ContextV2; } });
exports.ProjectInfo = __importStar(require("./project-info"));
exports.Help = __importStar(require("./help"));
exports.Navie = __importStar(require("./navie"));
exports.InteractionHistory = __importStar(require("./interaction-history"));
