// @appland/navie/dist/services/classification-service.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __asyncValues = (this && this.__asyncValues) || function (o) {
    if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
    var m = o[Symbol.asyncIterator], i;
    return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
    function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
    function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
};
Object.defineProperty(exports, "__esModule", { value: true });
const openai_1 = require("@langchain/openai");
const interaction_history_1 = require("../interaction-history");
const SYSTEM_PROMPT = `**Question classifier**

A software developer is asking a question about a project. Your task is to classify the question into one or more categories.

# About you**

You are part of a software system called Navie which is created and maintained by AppMap Inc, and is
available to AppMap users as a service. AppMap is a tool that helps developers understand, maintain and improve their codebases.
It works by running AI code analysis on code snippets, sequence diagrams, HTTP server and client requests, exceptions, log messages,
and database queries.

## About the user

The user is a software developer who is working on a project. They are asking a question about the project.

## Classification categories

Your task is to assign a likelihood to each of the following categories:

- **greeting**: The developer is saying hello to the assistant.
- **help-with-appmap**: Help using the AppMap product.
- **architecture**: The developer is asking about the architecture of the project.
- **feature**: The developer is asking for an explanation of how a specific feature of the project works.
- **overview**: The developer is asking a high-level question about the structure, purpose,
  functionality or intent of the project.
- **troubleshoot**: The developer is asking for help troubleshooting an issue.
- **explain**: The developer is asking for an explanation of a the code.
- **generate-code**: The developer is asking to generate code for a specific task.
- **generate-diagram**: The developer is asking for a diagram of their code.

## Classification scores

Each classification category is assigned one of the following likelihoods:

- **high**: The question is very likely to be of this type.
- **medium**: The question is somewhat likely to be of this type.
- **low**: The question is unlikely to be of this type.

**Response**

Respond with the likelihood of each question type. Question types with "low" likelihood may
be omitted.

**Examples**

Some examples of questions and their classifications are:

\`\`\`yaml
- question: Hi
  answer:
    greeting: high
    help-with-appmap: low
    architecture: low
    feature: low
    overview: medium
    troubleshoot: low
    explain: low
    generate-code: low
    generate-diagram: low

- question: How do I record AppMap data of my Spring app?
  answer:
    help-with-appmap: high
    architecture: low
    feature: low
    overview: low
    troubleshoot: low
    explain: low
    generate-code: low
    generate-diagram: low

- question: How do I reduce the amount of data in my AppMaps?
  answer:
    help-with-appmap: high
    overview: low
    troubleshoot: low
    explain: low
    generate-code: medium
    generate-diagram: low

- question: How does the project work?
  answer:
    help-with-appmap: low
    architecture: high
    feature: low
    overview: high
    troubleshoot: low
    explain: low
    generate-code: low
    generate-diagram: low

- question: Generate a form and controller to update the user profile
  answer:
    help-with-appmap: low
    architecture: medium
    feature: high
    overview: low
    explain: low
    generate-code: high
    generate-diagram: low

- question: Why am I getting a 500 error?
  answer:
    help-with-appmap: low
    architecture: low
    feature: low
    overview: low
    troubleshoot: high
    explain: medium
    generate-code: low
    generate-diagram: low

- question: Generate a diagram of the user profile feature
  answer:
    help-with-appmap: medium
    architecture: high
    feature: high
    overview: low
    troubleshoot: low
    explain: low
    generate-code: low
    generate-diagram: high
\`\`\`

`;
class ClassificationService {
    constructor(interactionHistory, modelName, temperature) {
        this.interactionHistory = interactionHistory;
        this.modelName = modelName;
        this.temperature = temperature;
    }
    classifyQuestion(question, chatHistory) {
        var _a, e_1, _b, _c;
        return __awaiter(this, void 0, void 0, function* () {
            const allQuestions = [
                ...(chatHistory || [])
                    .filter((message) => message.role === 'user')
                    .map((message) => message.content),
                question,
            ]
                .filter(Boolean)
                .join('\n\n');
            const openAI = new openai_1.ChatOpenAI({
                modelName: this.modelName,
                temperature: this.temperature,
            });
            const messages = [
                {
                    content: SYSTEM_PROMPT,
                    role: 'system',
                },
                {
                    content: allQuestions,
                    role: 'user',
                },
            ];
            // eslint-disable-next-line no-await-in-loop
            const response = yield openAI.completionWithRetry({
                messages,
                model: openAI.modelName,
                stream: true,
            });
            const tokens = Array();
            try {
                // eslint-disable-next-line no-await-in-loop
                for (var _d = true, response_1 = __asyncValues(response), response_1_1; response_1_1 = yield response_1.next(), _a = response_1_1.done, !_a;) {
                    _c = response_1_1.value;
                    _d = false;
                    try {
                        const token = _c;
                        tokens.push(token.choices.map((choice) => choice.delta.content).join(''));
                    }
                    finally {
                        _d = true;
                    }
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (!_d && !_a && (_b = response_1.return)) yield _b.call(response_1);
                }
                finally { if (e_1) throw e_1.error; }
            }
            const rawTerms = tokens.join('');
            const lines = rawTerms.split('\n');
            const classification = lines
                .map((line) => {
                if (!line.trim())
                    return null;
                const match = line.match(/([\w-]+)\s*:\s*(\w+)/);
                if (!match)
                    return null;
                // Sometimes the question is classified as "question" which is not a valid classification
                if (match[1] === 'question')
                    return null;
                return {
                    name: match[1],
                    weight: match[2],
                };
            })
                .filter((item) => item);
            const labels = classification;
            this.interactionHistory.addEvent(new interaction_history_1.ClassificationEvent(labels));
            return labels;
        });
    }
}
exports.default = ClassificationService;
