// @appland/appmap/built/cmds/open/showAppMap.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const node_process_1 = __importDefault(require("node:process"));
const child_process_1 = require("child_process");
const userInteraction_1 = __importDefault(require("../userInteraction"));
const openers_1 = require("./openers");
async function showAppMap(appMapFile) {
    function open() {
        if (node_process_1.default.platform === 'darwin' && node_process_1.default.env.TERM_PROGRAM === 'vscode')
            try {
                (0, child_process_1.spawn)('code', [(0, openers_1.abspath)(appMapFile)]);
                return;
            }
            catch (_a) {
                // fallthrough
            }
        return (0, openers_1.openInBrowser)(appMapFile, false);
    }
    await open();
    const { action } = await userInteraction_1.default.prompt({
        type: 'list',
        name: 'action',
        message: 'What would you like to do next:',
        choices: [`Reopen ${appMapFile}`, 'Exit'],
    });
    if (action.startsWith('Reopen')) {
        await showAppMap(appMapFile);
    }
}
exports.default = showAppMap;
